[FLINK-15999] Extract “Concepts” material from API/Library sections and start proper concepts section

aljoscha opened the pull request:
This is the second phase of https://cwiki.apache.org/confluence/display/FLINK/FLIP-42%3A+Rework+Flink+Documentation.

The first few comments are preliminary refactorings of the doc that are also in separate PRs. You can also review them here if you want.

The commits describe how the sections are moved around and when they are adapted/fixed. The general idea is that all concept-y material is extracted from individual sections and moved to the concepts section. On the way, some duplicate material was dropped and I fixed outdated references/naming.

I also changed all links to use the `{{ site.baseurl }}{% link dev/bar/baz/doc.md %}` format which has detection of missing links built in. 

I added `TODO: ...` entries in various places in the concepts sections to mark what still needs to be done. I think leaving them in for now is better than what we had before as a concepts section, I would create Jira Issues for those and try and spread the work. At the latest before the next release those should be removed.

Reviewing this PR is probably easiest by checking it out and reading the concepts sections.

knaufk commented on docs/concepts/stateful-stream-processing.md:
```diff
@@ -0,0 +1,412 @@
+support use cases where some data coming from one stream is required to be
+broadcasted to all downstream tasks, where it is stored locally and is used to
+process all incoming elements on the other stream. As an example where
+broadcast state can emerge as a natural fit, one can imagine a low-throughput
```
"As an example ... one can imagine" is a bit long-winded introduction into an example ;)

knaufk commented on docs/concepts/stateful-stream-processing.md:
```diff
@@ -0,0 +1,412 @@
+broadcasted to all downstream tasks, where it is stored locally and is used to
+process all incoming elements on the other stream. As an example where
+broadcast state can emerge as a natural fit, one can imagine a low-throughput
+stream containing a set of rules which we want to evaluate against all elements
```
Style guide says "use 'you', never 'we'".

knaufk commented on docs/concepts/stateful-stream-processing.md:
```diff
@@ -0,0 +1,412 @@
+ 1. it has a map format,
+ 2. it is only available to specific operators that have as inputs a
+    *broadcasted* stream and a *non-broadcasted* one, and
+ 3. such an operator can have *multiple broadcast states* with different names.
```
I think, the third bullet point is too detailed.

knaufk commented on docs/concepts/stateful-stream-processing.md:
```diff
@@ -0,0 +1,412 @@
+## State Persistence
+
+Flink implements fault tolerance using a combination of **stream replay** and
+**checkpointing**. A checkpoint is related to a specific point in each of the
```
related is very unprecise

aljoscha replied:
I'll change it to `marks a specific point`.

knaufk commented on docs/concepts/stateful-stream-processing.md:
```diff
@@ -0,0 +1,412 @@
+input streams along with the corresponding state for each of the operators. A
+streaming dataflow can be resumed from a checkpoint while maintaining
+consistency *(exactly-once processing semantics)* by restoring the state of the
+operators and replaying the events from the point of the checkpoint.
```
```suggestion
operators and replaying the records from the point of the checkpoint.
```
glossary

StephanEwen commented:
Thanks, this is a great first step!

Is the plan to merge this with the TODO comments? (I am personally fine with that)
Or is that a base PR on which you want to base the next ones?

A few minor suggestions:
  - I would remove the "Programming Model (outdated)" section.
  - Instead, having a staring page (maybe for concepts, maybe for the docs as a whole) that show the stack (SQL/Table API / DataStream API / StateFun, those on top of the common runtime)

  - I would drop operator state from the introduction. Instead I would mention it in the `DataStream API` as something that is mainly targeted towards connectors.

  - What do you think about removing ingestion time from the "timely" docs? We could change it to just a short mention under "event time", or mention it not at all.

  - I would also remove the "Asynchronous State Snapshots" section and simply have a comment somewhere that all persistence operations are asynchronous.

knaufk commented on docs/concepts/stateful-stream-processing.md:
```diff
@@ -0,0 +1,412 @@
+operators and replaying the events from the point of the checkpoint.
+
+The checkpoint interval is a means of trading off the overhead of fault
+tolerance during execution with the recovery time (the number of events that
```
```suggestion
tolerance during execution with the recovery time (the number of records that
```
glossary

knaufk commented on docs/concepts/stateful-stream-processing.md:
```diff
@@ -0,0 +1,412 @@
+streaming data flow. For streaming applications with small state, these
+snapshots are very light-weight and can be drawn frequently without much impact
+on performance.  The state of the streaming applications is stored at a
+configurable place (such as the master node, or HDFS).
```
usually in a distributed filesystem?

knaufk commented on docs/concepts/stateful-stream-processing.md:
```diff
@@ -0,0 +1,412 @@
+operators and resets them to the latest successful checkpoint. The input
+streams are reset to the point of the state snapshot. Any records that are
+processed as part of the restarted parallel dataflow are guaranteed to not have
+been part of the previously checkpointed state.
```
```suggestion
effected the previously checkpointed state.
``

knaufk commented on docs/concepts/stateful-stream-processing.md:
ability twice

knaufk commented on docs/concepts/stateful-stream-processing.md:
I usually use snapshots to subsume checkpoints and savepoints.

knaufk commented on docs/concepts/stateful-stream-processing.md:
```diff
@@ -0,0 +1,412 @@
+processed as part of the restarted parallel dataflow are guaranteed to not have
+been part of the previously checkpointed state.
+
+{% info Note %} By default, checkpointing is disabled. See [Checkpointing]({{
```
style guide says use `<div class="alert alert-info"> // Info Message </div>` for notes

aljoscha replied:
Most of the documentation uses the `{% info/warn %}` so I will not change it here. We can do that throughout the documentation in the future.

knaufk commented on docs/concepts/stateful-stream-processing.md:
Feels like Barriers, Snapshotting Operator State, Asynchronous State Snasphots and Recovery are sub-section of checkpointing.

aljoscha replied:
you're right, I'll change that. 👌

knaufk commented on docs/concepts/stateful-stream-processing.md:
I agree with Stephan that this quite implementation detaily.

knaufk commented on docs/concepts/stateful-stream-processing.md:
all streaming programs, right?

knaufk replied:
hence all programs that use checkpointing

aljoscha replied:
yes 👌

knaufk commented on docs/concepts/stateful-stream-processing.md:
Everything starting from "During" is not about savepoints and can be removed imho.

aljoscha replied:
👌

knaufk commented on docs/concepts/stateful-stream-processing.md:
```suggestion
Savepoints are similar to checkpoints except that they are
```

knaufk commented on docs/concepts/stateful-stream-processing.md:
How savepoint are created is not conceptual content, but operational.

knaufk commented on docs/concepts/stateful-stream-processing.md:
Are iterations about state?

aljoscha replied:
in a way they are because iterations repeatedly join a stream with the existing state

knaufk commented on docs/concepts/flink-architecture.md:
```diff
@@ -0,0 +1,140 @@
+  - The *Flink Master* coordinates the distributed execution. It schedules
+    tasks, coordinates checkpoints, coordinates recovery on failures, etc.
+
+    There is always at least one *Flink Master*. A high-availability setup will
```
```suggestion
    There is always at least one *Flink Master*. A high-availability setup might
```

knaufk commented on docs/concepts/flink-architecture.md:
```diff
@@ -0,0 +1,140 @@
+    the others are *standby*.
+
+  - The *TaskManagers* (also called *workers*) execute the *tasks* (or more
+    specifically, the subtasks) of a dataflow, and buffer and exchange the data
```
according to the glossary "(or more specifically, the subtasks)" does not make sense.

knaufk commented on docs/concepts/flink-architecture.md:
```diff
@@ -0,0 +1,140 @@
+The sample dataflow in the figure below is executed with five subtasks, and
+hence with five parallel threads.
+
+<img src="{{ site.baseurl }}/fig/tasks_chains.svg" alt="Operator chaining into Tasks" class="offset" width="80%" />
```
Top: Logical Graph, there are no tasks in the logical graph
Bottom: Physical Graph

knaufk commented on docs/concepts/flink-architecture.md:
```diff
@@ -0,0 +1,140 @@
+subtasks in separate threads.  To control how many tasks a worker accepts, a
+worker has so called **task slots** (at least one).
+
+Each *task slot* represents a fixed subset of resources of the TaskManager. A
```
There is already quite some operational content mixed in here. Could be shorter in the Concepts section in my opinion.

knaufk commented on docs/concepts/flink-architecture.md:
```diff
@@ -0,0 +1,140 @@
+<img src="{{ site.baseurl }}/fig/tasks_slots.svg" alt="A TaskManager with Task Slots and Tasks" class="offset" width="80%" />
+
+By default, Flink allows subtasks to share slots even if they are subtasks of
+different tasks, so long as they are from the same job. The result is that one
```
```suggestion
different operators, so long as they are from the same job. The result is that one
```

knaufk commented on docs/concepts/flink-architecture.md:
```diff
@@ -0,0 +1,140 @@
+
+<img src="{{ site.baseurl }}/fig/slot_sharing.svg" alt="TaskManagers with shared Task Slots" class="offset" width="80%" />
+
+The APIs also include a *[resource group]({{ site.baseurl }}{% link
```
From here onwards, this is not conceptual anymore in my opinion.

knaufk commented on docs/concepts/timely-stream-processing.md:
```diff
@@ -0,0 +1,237 @@
+
+<img src="{{ site.baseurl }}/fig/parallel_streams_watermarks.svg" alt="Parallel data streams and operators with events and watermarks" class="center" width="80%" />
+
+Note that the Kafka source supports per-partition watermarking, which you can
```
too specific

knaufk commented on docs/concepts/timely-stream-processing.md:
```diff
@@ -0,0 +1,237 @@
+
+<img src="{{ site.baseurl }}/fig/windows.svg" alt="Time- and Count Windows" class="offset" width="80%" />
+
+More window examples can be found in this [blog
```
Please check out this blog post for additional examples of windows or take a look a window documentation of the DataStream API.

knaufk commented on docs/dev/stream/state/index.md:
```diff
@@ -25,23 +25,10 @@ specific language governing permissions and limitations
-The [queryable state](queryable_state.html) feature of Flink allows you to access state from outside of Flink during runtime.
-
-When working with state, it might also be useful to read about [Flink's state backends]({{ site.baseurl }}/ops/state/state_backends.html). Flink provides different state backends that specify how and where state is stored. State can be located on Java's heap or off-heap. Depending on your state backend, Flink can also *manage* the state for the application, meaning Flink deals with the memory man ...
+In this section you will learn about the stateful abstractions that Flink
```
are abstractions stateful?

aljoscha replied:
`learn about the APIs that Flink provides for writing stateful programs`?

knaufk commented on docs/dev/stream/state/state.md:
```diff
@@ -22,66 +22,17 @@ specific language governing permissions and limitations
 -->
 
-This document explains how to use Flink's state abstractions when developing an application.
+In this section you will learn about the stateful abstractions that Flink
```
see above

aljoscha commented:
Thanks a lot for the reviews!

Regarding TODOs about extending sections: I would prefer to leave them in, merge this PR rather quickly, and create follow-up Jira Issues. This reworking is already big enough as it is and if we try to to the one big everything-solution I'm afraid this will drag on quite a bit. Incremental improvements should work well here. I will therefore also not address all comments, for example about figures or longer sections that I would have to newly create or that would take too much time now. I think this change is already an improvement on the previous state, as is.

@StephanEwen I addressed your comments. Some of them I already had as TODOs in the document, it's good to see them confirmed and I not did the actual changed.

@knaufk I have yet to go through your suggestions and apply them.

aljoscha commented on docs/concepts/stateful-stream-processing.md:
```diff
@@ -0,0 +1,412 @@
+
+<img src="{{ site.baseurl }}/fig/state_partitioning.svg" alt="State and Partitioning" class="offset" width="50%" />
+
+Keyed State is further organized into so-called *Key Groups*. Key Groups are
```
It can be important to understand the behaviour of Flink in certain situations, but I'm happy to remove this.

aljoscha commented:
I now also addressed @knaufk's comments. I did not address comments that "go deeper":
 - I didn't update figures, except some very simple changes
 - I didn't change how we refer to *dataflow*/*graph*/*logical graph* and the like because I think that requires changes in more parts
 - I didn't change how we refer to the distribution patterns of streams
 - I didn't change how we refer to *task*/*subtask*/*operator*/*operation*/*transformation*

I don't think that we shouldn't do this at all but I don't have the capacity to do all of this in one go. IMO it's better to first get all the basic reorganisation in place and then to more passes over the doc to improve it continuously.

knaufk commented:
@aljoscha I see. How can we make sure, these comments are not lost. Directly create sub-tasks?

aljoscha commented:
@knaufk I think yes, on the subtasks.

aljoscha commented:
Thanks for the reviews! I created new subtasks for the `TODOs` and for things that were mentioned in the review but which I didn't address.
